### PR TITLE
Support multiple subscriptions for the Insight device

### LINKED ISSUE #115: Allow pywemo to subscribe to multiple services
At present the pywemo subscription code only subscribes to the `basicevent` services.

This updates on the most important events like (on / off) - but doesn't work for everything (eg for wemo insight switches - it doesn't update when the power consumption is changed (unless it also triggers a change in or out of standby - which does trigger a `basicevent`).

For this reason I had to leave polling turned on in Home Assistant for switches (@sqldiablo - you might want to check this still works in your recent HA refactor) so that insight parameters would still be updated correctly.

Insight parameter changes are reported by a subscription to `insight`. I expect a number of other wemo devices that report their state via other services such as  https://github.com/pavoni/pywemo/issues/6).

The ghost of my attempt to update this is here
https://github.com/pavoni/pywemo/blob/master/pywemo/subscribe.py#L119

The current re-subscribe code is set up to handle a single subscription per device code so will need to be refactored to support this.

cc @sqldiablo in case you are interested in taking a look at this.

### PR DESCRIPTION
**Pull request by esev**
## Description:

Support multiple subscriptions for the Insight device.

The original subscription logic stored the scheduler Event instances in the `self._events` dictionary, keyed by the device serial number. This only permitted one subscription (one scheduler Event) per device.

Original: `self._events[device.serialnumber] = self._sched.enter(...)`

This PR adds an additional level of dictionary. `self._events` is still keyed by serialnumber, but rather than a single Event instance, it now contains a dict with multiple Event instances. Each individual Event is keyed by the function used to fetch the URL for the subscription.

New: `self._events[device.serialnumber][url_fn] = self._sched.enter(...)`

Two new functions are added, each for a specific subscription URL.

```python
def _basic_event_subscription_url(device: Device) -> str:
    """Return the basic event subscription URL."""
    return device.basicevent.eventSubURL


def _insight_event_subscription_url(device: Device) -> str:
    """Return the insight event subscription URL."""
    return device.insight.eventSubURL
```

These functions are the keys for the second-level dictionary. They are also used to lookup the appropriate device URL to maintain the subscription.

It was important in this PR to not store the URL of the device directly, but rather read it from the device instance each time. This is necessary to support probing/rediscovery of a device when its IP/port changes. PR #209 made the mistake of using the URL as the second-level dictionary key and storing it in the Event arguments. #209 was flawed because the URL/host/port never changed when the device was rediscovered. Using the url_fn avoids that mistake.

**Related issue (if applicable):** fixes #115

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] There is no commented out code in this PR.

### CODE REVIEW

File: tests/vcr/tests.test_subscribe/Test_SubscriptionRegistry.test_register_unregister.yaml [hunk 1]
```diff
@@ -39,4 +39,44 @@ interactions:
     status:
       code: 200
       message: OK
+- request:
```
Comment (Spectre5) [inline tests/vcr/tests.test_subscribe/Test_SubscriptionRegistry.test_register_unregister.yaml:42]: Empty body and empty response body, is that right?

**Comment by esev** [comment]
@Spectre5 when you have a chance, could you take a look at the build failure? I think it isn't related to this PR, but I could be mistaken.

```
Running on Github Actions but GITHUB_TOKEN is not set. Add "env: GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}" to your step config.
Traceback (most recent call last):
  File "/home/runner/work/pywemo/pywemo/.venv/lib/python3.9/site-packages/coveralls/cli.py", line 61, in main
    coverallz = Coveralls(token_required,
  File "/home/runner/work/pywemo/pywemo/.venv/lib/python3.9/site-packages/coveralls/api.py", line 47, in __init__
    self.ensure_token()
  File "/home/runner/work/pywemo/pywemo/.venv/lib/python3.9/site-packages/coveralls/api.py", line 54, in ensure_token
    raise CoverallsException(
coveralls.exception.CoverallsException: Running on Github Actions but GITHUB_TOKEN is not set. Add "env: GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}" to your step config.
Error: Process completed with exit code 1.
```

File: tests/vcr/tests.test_subscribe/Test_SubscriptionRegistry.test_register_unregister.yaml (hunk 1 above)
Comment (esev) [reply tests/vcr/tests.test_subscribe/Test_SubscriptionRegistry.test_register_unregister.yaml:42]: That is correct. These requests use only the headers to pass values.

More details can be found in the [UPnP Device Architecture](http://www.upnp.org/specs/arch/UPnP-arch-DeviceArchitecture-v1.0-20080424.pdf) spec starting around page 64.

**Comment by Spectre5** [comment]
Definitely not related to your PR, I'll add that in a new PR right now.

**Comment by Spectre5** [comment]
ok, rebase on master and try again

**Comment by esev** [comment]
@Spectre5 you likely already saw the new build failure. No rush on this. I'm more than happy to rebase/push as many times as necessary so everything can be worked out.

**Comment by Spectre5** [comment]
I suspect that this PR might need to be closed and re-opened or re-created.  I found in one of the various sites I read that after fixing this 422 error (which I hope I have now in master), that they could only get their current PR's to respond correctly after closing and re-opening or re-creating it.

**Comment by esev** [comment]
> I suspect that this PR might need to be closed and re-opened or re-created.

Nice find! Closing and reopening solved it. Thanks!

**Comment by Spectre5** [comment]
whew :sweat_smile: haha

**Comment by Spectre5** [comment]
I haven't really looked at this closely myself yet, what's the status/verdict on it?

**Comment by esev** [comment]
It is working for me. I'd prefer to have a 2nd look at changes like this before merging, but if you think it's okay I can merge it.

**Comment by Spectre5** [comment]
Ok I can take a look at it for logic, just can't physically test it since I don't have an insight!

**Comment by pavoni** [comment]
I'll give it a try - I had a quick look at the code - which looks good. But I'll also run it a little to double check.

**Comment by Spectre5** [comment]
We should add a test_insight file too, to test the methods in the Insight class.  Could tag it on to this PR or make a new one.

**Comment by pavoni** [comment]
This looks good - and works well in my setup.

Are we all happy to merge?

**Comment by esev** [comment]
It has been working well in my HA too. I haven't observed any issues. I'm good to merge.

It sure catches a lot more state changes :)
Screenshot 2021-02-04 at 9 04 15 AM

**Comment by pavoni** [comment]
Thanks @esev, let's check @Spectre5 is happy to, and then I'll merge and release.

It looks like HA benefits quite a bit - even without changes to the HA wrapper - so once we've merged shall I do an HA version bump. It'll be a simple PR so will likely be merged easily (the rules PR is a lot more complex so might take more time to review)?

**Comment by pavoni** [comment]
And BTW thanks for this - as you know I tried (and failed) to do this myself a few years back!

**Comment by esev** [comment]
> once we've merged shall I do an HA version bump. It'll be a simple PR so will likely be merged easily (the rules PR is a lot more complex so might take more time to review)?

Thank you @pavoni! I agree, version bumps go quickly (https://github.com/home-assistant/core/pull/45501), feature requests go slower (https://github.com/home-assistant/core/pull/45503).

If either of you have time to review https://github.com/home-assistant/core/pull/45503, maybe that will speed it along?

**Comment by esev** [comment]
> once we've merged shall I do an HA version bump.

Actually I misread that as "I shall do". :) I'll take care of it.

**Review by Spectre5 — APPROVED** [review]

**Comment by Spectre5** [comment]
looks good to me.

**Pull request merged**